ouroboros-network/src/Ouroboros/Network/Block.hs

```diff
@@ -114,21 +116,26 @@ castHash (BlockHash b) = BlockHash b
 -- as a check, or in some contexts it disambiguates blocks from different forks
 -- that were in the same slot.
 --
-data Point block = Point {
-       pointSlot :: SlotNo,
-       pointHash :: ChainHash block
-     }
-   deriving (Eq, Ord, Show)
+data Point block = GenesisPoint | Point !SlotNo !(HeaderHash block)
```

Comment:
This ought to be factored into

```Haskell
data Point t = Origin | Point t

data BlockPoint slot hash = BlockPoint slot hash
```

because `BlockPoint` does find uses: any case in which you want to identify an actual block, without including the origin. For instance, in your block fetch stuff, you don't want `Point -> Bool`, you want `BlockPoint -> Bool`, because obviously you have already fetched the origin (there is no block here).

Reply:
Another concrete example: in the forthcoming sqlite index implementation that supports rollbacks, I use a `Point (BlockPoint IndexSlot HeaderHash)` for rollbacks (includes the origin) but a `BlockPoint IndexSlot HeaderHash` for rollforwards, because you cannot roll forward to the origin.